Checksum inputdata

jedwards4b opened the pull request:
Add checksum support for inputdata:  When files are downloaded the data is checksummed against the file inputdata_chksum.dat which has a format:
```  
 df9a88f7dd33ea7451307e2b3b68cea5             1445  eaton            1661657404  Feb  20   2010 inputdata/atm/cam/chem/1850-2000_emis/IPCC_emissions_aircraft_BC_1850-2000_1.9x2.5.c090729.nc
```
the first and last fields read into a hash and as files are downloaded the file check sum is compared to the hash in the file.  


Test suite: by hand, scripts_regression_tests.py
Test baseline: 
Test namelist changes: 
Test status: bit for bit
Fixes #2910 

User interface changes?: adds a --chksum option to check_input_data, files downloaded are always checked, the option is to check files already on local disk for a given case

Update gh-pages html (Y/N)?:

Code review:

ekluzek commented:
This would also help in finding files that have zero size or aren't available on svn inputdata, so it wouldn't create a local file of zero size. This would help prevent problems like:

https://github.com/ESCOMP/ctsm/issues/646

where a zero size file is created, because it's not in svn.

jedwards4b commented:
@ekluzek if a file is not on the server an error is printed and a zero length file is no longer produced.
running ./check_input_data --chksum should revile any relevant files in DIN_LOC_ROOT which do not
match the chksum file.

billsacks commented on scripts/lib/CIME/case/check_input_data.py:
```diff
@@ -6,9 +6,45 @@
+    """
+    Return True if successfully downloaded
+    """
+    rel_path = "../inputdata_checksum.dat"
```
Is it a problem that there is only one checksum file despite potentially having multiple inputdata servers? I would imagine having a separate checksum file for each server, with different file names distinguishing them. But maybe I'm missing something here.

Maybe this isn't a huge deal, because (if I understand the operation of this code correctly), a given call to `check_input_data` only works with a single remote server, and so would create a new `inputdata_checksum.dat` for that server. But it seems like this could still create some confusion, especially if there are any problems.

I don't feel strongly that this needs to be fixed, but wanted to run this by you.

jedwards4b replied:
I agree that this could be a problem, I've dealt with it as follows:  The server checksum file is now listed in config_inputdata.xml and the client checksum file is a unique merge of server checksum files.   If the same file has different checksums on different servers an error will result.

billsacks replied:
Your solution seems good - thanks.

billsacks commented on scripts/lib/CIME/case/check_input_data.py:
```diff
@@ -6,9 +6,45 @@
+    rel_path = "../inputdata_checksum.dat"
+    full_path = os.path.join(input_data_root, "inputdata_checksum.dat")
+    protocol = type(server).__name__
+    if protocol == 'svn':
```
It seems like this logic is specific to our current svn inputdata server, and should be encoded in the xml defining the different servers rather than generally skipping this for any svn server.

jedwards4b replied:
you are correct, i've added the file name to the config_inputdata.xml

billsacks commented on scripts/lib/CIME/case/check_input_data.py:
```diff
@@ -6,9 +6,45 @@
+    if protocol == 'svn':
+        # svn server does not have this file
+        return False
+    logging.info("Trying to download file: '{}' to path '{}' using {} protocal.".format(rel_path, full_path, protocol))
```
Spelling: "protocal" should be "protocol"

billsacks commented on scripts/lib/CIME/case/check_input_data.py:
```diff
@@ -6,9 +6,45 @@
+        else:
+            if os.path.isfile(tmpfile):
+                os.rename(tmpfile, full_path)
+                logger.warning("Could not automatically download file "+full_path+
```
Connected with some of my above comments: it seems like this warning is misleading for two reasons:

(1) Based on what you told me in person, it sounds like it's acceptable for some servers to not have this file, so maybe this shouldn't generate a warning? I wonder if each server's xml entry should say whether that file has a checksum or not; then we'll skip trying to download it if it doesn't have one, and just try to download it if it is supposed to have one (generating this warning if it isn't found).

(2) The comment about "Restoring existing version" is a bit misleading because it may be restoring a file for a completely different server right now.

billsacks commented on scripts/lib/CIME/case/check_input_data.py:
```diff
@@ -6,9 +6,45 @@
+                           " Restoring existing version.")
+            else:
+                logger.warning("Could not automatically download file "+full_path+
+                           "download from ftp:ftp.cgd.ucar.edu/cesm/inputdata_chksum.dat")
```
Shouldn't this hard-coded path be replaced with the actual remote path for this particular server?

billsacks commented on scripts/lib/CIME/case/check_input_data.py:
```diff
@@ -37,15 +73,16 @@ def _download_if_in_repo(server, input_data_root, rel_path, isdirectory=False):
 ###############################################################################
-def check_all_input_data(self, protocol=None, address=None, input_data_root=None, data_list_dir="Buildconf", download=True):
+def check_all_input_data(self, protocol=None, address=None, input_data_root=None, data_list_dir="Buildconf",
+                         download=True, chksum=False):
```
Please add documentation for the new chksum argument. (Ideally there would be documentation for all arguments.)

billsacks replied:
This comment and my other similar comment about argument documentation are still unresolved: this and other functions could use more documentation of their arguments.

billsacks commented on scripts/lib/CIME/case/check_input_data.py:
```diff
@@ -129,7 +166,8 @@ def stage_refcase(self, input_data_root=None, data_list_dir=None):
 
-def check_input_data(case, protocol="svn", address=None, input_data_root=None, data_list_dir="Buildconf", download=False):
+def check_input_data(case, protocol="svn", address=None, input_data_root=None, data_list_dir="Buildconf",
+                     download=False, chksum=False, user=None, passwd=None):
```
Please add documentation for the new chksum argument. (Ideally there would be documentation for all arguments.)

billsacks commented on scripts/lib/CIME/case/check_input_data.py:
```diff
@@ -205,13 +244,57 @@ def check_input_data(case, protocol="svn", address=None, input_data_root=None, d
+
+
+def verify_chksum(input_data_root, checksumfile, filename):
+    if not chksum_hash:
```
Similar to my above comment about only having one checksum file for all servers: I think it's a problem that you only have a single chksum_hash variable for all servers. My understanding of the logic here is that you'll read the checksum file for the first server, but then for each subsequent server, you won't reset chksum_hash, but instead will use the variable from the first server. If my understanding is right, then this is a problem.

billsacks commented on scripts/lib/CIME/case/check_input_data.py:
```diff
@@ -205,13 +244,57 @@ def check_input_data(case, protocol="svn", address=None, input_data_root=None, d
+
+    chksum = md5(os.path.join(input_data_root, filename))
+    if chksum_hash:
+        expect(filename in chksum_hash.keys() and chksum == chksum_hash[filename],
```
It seems like there are a few problems with expecting the current file to be in chksum_hash:

1. Tied in to my above comments: let's say we just downloaded a file from server B, but chksum_hash was set based on server A; then this new file would be missing from chksum_hash.

2. When running `check_input_data` with the new `--chksum` argument: My impression is that you'll end up trying to do this checksum for each server, but the file only exists on one server, so wouldn't the expect fail for all of the other servers? This case also has the problem of (1). In addition, due to the problem of using the same file name for all servers, this case has the problem that the checksum file on disk may never be for the server this file resides on (if no files need to be downloaded from that server), because it looks like existing files don't force a download of the checksum file from the server.

3. Can we assume that the checksum file on the server is updated near-instantaneously whenever a file is added to the server? If not, we could have a situation where you upload a file from one machine, then a few minutes later go to download it on a different machine, and get an error because the new file is not yet in the server's checksum file.

billsacks replied:
In addition, if you do keep the check of filename in chksum_hash.keys(), you should make this a separate check, rather than checking both conditions at once: These two issues (missing file name in the chksum file and mismatching chksum) are fundamentally different problems, and it's confusing to have them reported with the same error.

billsacks commented on scripts/lib/CIME/case/check_input_data.py:
```diff
@@ -205,13 +244,57 @@ def check_input_data(case, protocol="svn", address=None, input_data_root=None, d
+        hashfile = os.path.join(input_data_root,checksumfile)
+        if not os.path.isfile(hashfile):
+            expect(False, "Failed to find or download file "+hashfile+
+                   " download from ftp:ftp.cgd.ucar.edu/cesm/inputdata_chksum.dat")
```
Again, shouldn't this path be dynamic?

jedwards4b replied:
It turns out that I should remove this because the the file is now reformatted after download

billsacks commented on scripts/lib/CIME/case/check_input_data.py:
```diff
@@ -205,13 +244,57 @@ def check_input_data(case, protocol="svn", address=None, input_data_root=None, d
+                lsplit = line.split()
+                if len(lsplit) < 8:
+                    continue
+                fname = (lsplit[7])[10:]
```
I'm uncomfortable with the hard-coded [10:]. Can you instead just strip the leading 'inputdata/' from this string?

billsacks commented on scripts/lib/CIME/case/check_input_data.py:
```diff
@@ -205,13 +244,57 @@ def check_input_data(case, protocol="svn", address=None, input_data_root=None, d
+            for line in lines:
+                lsplit = line.split()
+                if len(lsplit) < 8:
+                    continue
```
Would it be feasible to require a header line on the checksum file, so that you could confirm that indeed the 1st column is the checksum and the 8th column is the file name?

I don't feel this is necessary, but it could help avoid some hard-to-debug errors in the future.

jedwards4b replied:
I am working with an existing chksum file that Mark Moore generates - I didn't really want to make any request that he change the format of the file on his server.

billsacks commented:
In addition to my review comments: I'm uncomfortable with this coming to maint-5.6 so close to the CESM2.1.1 release. I'd prefer if we had more time to exercise this functionality before it made it into a release - either by waiting to bring this to maint-5.6 until after this branch is tagged for CESM2.1.1, or by bringing it into cime master first.

billsacks commented:
> or by bringing it into cime master first.

However, if you do rebase this to be off of master, I'd find it easiest if you addressed my review comments in this PR before closing the PR and opening a new one.

jedwards4b commented on scripts/lib/CIME/case/check_input_data.py:
```diff
@@ -6,9 +6,41 @@
+    """
+    rel_path = "../inputdata_checksum.dat"
+    full_path = os.path.join(input_data_root, "inputdata_checksum.dat")
+    logging.info("Trying to download file: '{}' to path '{}' using {} protocal.".format(rel_path, full_path, type(server).__name__))
```
fixed these

jedwards4b commented:
@billsacks I think that I've addressed your concerns - my understanding is that the 2.1.1 release won't be until at least June so plenty of time to put this on maint-5.6

jedwards4b commented:
I've also added a lockfile for DIN_LOC_ROOT to assure that more than one case or user is not trying to update at the same time.   It also gives a better error if permissions do not allow the user to write there.

billsacks commented:
@jedwards4b Thanks for addressing my comments. I haven't looked closely yet, but I'll try to do so soon (not sure if I'll have time today, though).

I wanted to reply to the file locking piece first, though: I hadn't thought of this issue, but you raise a good point about the problem if multiple cases are trying to do this at once. However, I'm not sure I like the solution of adding this top-level lock on the check_input_data call: it seems like this could very easily bite people in practice. I'm especially thinking of test suites, where multiple cases are working simultaneously... and when you multiply that by multiple users on a system, it seems almost inevitable that cases will fail sometimes due to this lock.

Can we think a little more about what purpose(s) this lock serves, and whether this can be done in a different way that would reduce lock-induced failures to a bare minimum? I'm thinking:

1. Due to the changes in this PR, we need a lock on the local checksum file, to ensure that two processes aren't trying to write this checksum file at the same time. But: is there actually much value in having the checksum file reside permanently in this inputdata location? Would things work just as well if the checksum file was written as a temporary file (e.g., via https://docs.python.org/2/library/tempfile.html)? You could save the name of the temporary file in a module-level variable so that a given process would only need to download the checksum file once per server, but this would avoid clashes between different processes. However, I forget the details of the logic, so let me know if this wouldn't work because you rely on the persistence of this checksum file for some reason.

2. Were you also thinking about the need for locking if two cases are trying to download the same file from the server at the same time? (This is no more of an issue in this PR than it would have been previously, but maybe you were thinking of this here, too?) Or would the OS already prevent this? If the OS wouldn't already prevent this, then can we do a per-file locking mechanism to just lock what we absolutely need to lock? i.e., in the directory where we're about to download a file, putting a FILENAME.lock file, then removing that file as soon as that file is downloaded.

jedwards4b commented:
@billsacks the --chksum option assumes that the inputdata_checksum.dat file is present so it does need to be there for this to work.   This option allows a user to chksum all of the existing local input files for validity and doesn't require the user to have write permission to inputdata.   

Most of the time we do not need to download new data for testing or running cases, but it is true that if  multiple cases in a test suite need to download this could cause failures.   I'll look into changing this to a per file lock as you suggest.

billsacks commented:
@jedwards4b thanks for changing this to a per-file lock. This addresses point (2) from my earlier comment. But my impression is that now there is no lock on the checksum files. It really feels to me like the best thing to do with the checksum files is to download them to a temporary location, even when running with `--chksum`. This doesn't have to be in the inputdata space (since, as you point out, the user may not have write access there): e.g., I seem to remember that python's tempfile module will create a temporary file/directory in some special, OS-specific temporary location on disk where you have write permission; or perhaps this checksum file could be put in the case or run directories.

jedwards4b commented on scripts/lib/CIME/case/check_input_data.py:
```diff
@@ -6,19 +6,84 @@
+    success = False
+    rel_path = chksum_file
+    full_path = os.path.join(input_data_root, local_chksum_file)
+    lockfile = _check_permissions_and_lock_inputdata_file(input_data_root, user)
```
@billsacks here is the lock on the checksum file.

jedwards4b replied:
The checksum file on local disk may be a combination of checksums from several servers and 
is available to any user once a user with write access creates it.  The --chksum option may currently be used without any access to the inputdata server or servers, it seems useful to me that you would be able to use this option when servers are not available.

billsacks replied:
Thanks for pointing that out. But I still stand by my feeling that this checksum file should be in a case-specific location, rather than having the same file shared between all cases and users on the system. The more I think about this, the more strongly I feel about it: I feel that having a single file that is potentially written by all cases, for all users, is asking for trouble.

That said, if you still disagree, can we get a 3rd opinion here - from @jgfouca or someone else?

billsacks replied:
(My response was meant in response to your first comment here, not your second. I'll think about your second comment for a few minutes.)

billsacks replied:
> The checksum file on local disk may be a combination of checksums from several servers and
> is available to any user once a user with write access creates it. The --chksum option may currently be used without any access to the inputdata server or servers, it seems useful to me that you would be able to use this option when servers are not available.

I see your point, but I feel that the value of being able to run this when servers are unavailable does not justify the race condition issues caused by sharing a single file amongst all users and cases.

billsacks commented on scripts/lib/CIME/case/check_input_data.py:
```diff
@@ -205,13 +279,80 @@ def check_input_data(case, protocol="svn", address=None, input_data_root=None, d
+            for line in lines:
+                fchksum, fname = line.split()
+                if fname in chksum_hash:
+                    expect(chksum_hash[fname] == fchksum, " Inconsistant hashes in chksum for file {}".format(fname))
```
This is a good error-check - thanks.

Minor spelling error: Inconsistant should be Inconsistent.

billsacks commented:
Regarding my main remaining concern: One thing I've been wondering is: Why is there a single file per server with the checksums for all files on that server, as opposed to having a separate checksum file for each file on the server? It seems like the latter would minimize the potential for race conditions, and I think would simplify this PR, but maybe there's some reason for the current design that I'm not seeing.

gold2718 commented:
@billsacks, I have been thinking about this question. I think the main reason for a single checksum file (as opposed to the usual practice of a checksum file for every downloadable file) is disk management. On Cheyenne, there are currently over 23K NetCDF files in inputdata. I think that many small (e.g., checksum) files is a real drain on a large glade system. @jedwards4b, does that match your reasoning?
For managed HPC centers like NWSC it would make sense to me to have a process to create and maintain a central file which is fairly stable over time while individual users should not be held up.
A temporary download of checksums to the case directory avoids deadlocks. An attempt can be made to add new checksums to the master file but if it is locked, just give up -- it will eventually get added.

jgfouca commented:
@billsacks , @jedwards4b , I'm taking another look at this PR.

My understanding of the checksumming algorithm is:
1. User requests a checksum
2. Retrieve the name of the checksum file for that server
3. A lock file is created in the input data area with the name from (2) + .lock
4. The file is downloaded, which contains a map of filename to checksum for every file on that server
4.A If the file already exists, it is renamed so it can be merged later
5. The file is looked-up and the checksum is validated

The problem I see is that steps 3, 4, and 4.A all touch shared files but none of the steps are atomic, so a wide variety of races is possible. I could maybe be inclined to tolerate some risk here, but I also think the current solution is more complicated than it needs to be, so I'm going to recommend a refactor.

I'm thinking why do we need to maintain a checksum dictionary locally at all? I'm sure we could come up with a solution where a shell command could be used to directly request the needed checksum without touching the file system. Something like:
```
wget $remote_checksum_dict_file | grep filename | cut ...
```

jedwards4b commented:
So there are two different workflows here:

1. check_input_data determines that not all input_data for a case is available in the local inputdata directory.  In this case if the chksum file is available on the server it is downloaded and merged with any preexisting chksum file.  This merged file is then used to checksum each file from the server as it is downloaded.   Each time a file is downloaded a filename.LOCK file is written prior to the download and deleted when the download is complete.  This assures that two check_input_data processes don't attempt to download the same file at the same time. 
  
2.  A user requests to checksum all of the inputdata for their case using the --chksum option.  In this case the existing chksum file in input_data_root is used and each local file is checksummed.  No access to any server occurs.  There is no write of any file.    

I've added documentation throughout the check_input_data.py

jedwards4b commented:
@goldy my actual reasoning was that the cgd server has a single existing checksum file with a given format, I chose to work with that file in that format rather than ask the data steward to change his practice.

jgfouca commented:
@jedwards4b , OK, that makes sense, but I think I'd still prefer if download_checksum returned a string of all the checksum data instead of interacting with the file system. For your case 2, I assume it's not a problem if we access the server? The download of the checksum data should be real quick and I don't think it's worth it to do complex things in CIME to avoid this download. The lockfile solution reduces the race window, but I promise you it's still there. It also introduces another potential bug of leaving the lock file around if the CIME process crashes/raises exception at the wrong time.

jedwards4b commented:
okay - I think that I can rework it a bit so that the checksum file is downloaded to the case rundir for each case - is that acceptable as a way to avoid race conditions?

jgfouca commented:
@jedwards4b , yes, but why even store it in a file?

jgfouca commented:
Oh, I suppose it allows you to use Server.get_file. I'm fine with that.

billsacks commented:
> okay - I think that I can rework it a bit so that the checksum file is downloaded to the case rundir for each case - is that acceptable as a way to avoid race conditions?

Yes, that sounds good - thanks a lot.

You have addressed my earlier, specific points, so this rework of the location of the checksum file is the only thing left from my review.

jedwards4b commented:
@jgfouca @billsacks Please have another look.  I've moved the chksum file into the case directory and download it each time instead of trying to maintain it locally.

jgfouca commented:
@jedwards4b , now that you're using the case's rundir as the download location, can't you get rid of all the lockfile stuff?

jedwards4b commented:
I'm still downloading data to the inputdata directory the lock file stuff is an improvement somewhat independent of the chksum.

jgfouca commented:
@jedwards4b , instead of locking the entire inputdata directory, I would recommend using the atomic write pattern using file renaming for the individual downloads:

1. User needs to download file x/y/z.nc
2. Download file to x/y/z.nc.$unique
3. rename x/y/z.nc.$unique to x/y/z.nc

Where $unique is some string to ensure the filename is unique. It could be process-id.

jedwards4b commented:
I am not locking the entire inputdata directory - the lock is on individual files.

jgfouca commented:
@jedwards4b , oh, that's better, but I still think atomic write is best.

jedwards4b commented:
atomic write patterns are already happening implicitly - all of the download methods do this internally - this lock is to prevent a second download of the same file while the first is underway.

jgfouca commented:
> atomic write patterns are already happening implicitly

Can you elaborate? I don't see where this is happening.

If we are already doing atomic writes, then no locking is necessary. The worst thing that could happen is a bit of wasted bandwidth.

jedwards4b commented:
We do need locking - if two concurrently running cases need the same file and see that the file is not in inputdata they will independent try to download it.  When a case starts to download a file it creates this lock so that another case will know not to.  We occasionally run our tests with the DIN_LOC_ROOT pointed to an empty directory - this assures that all of the needed input data is in the repository - but there are all sorts of conflicts that occur because multiple tests are attempting to simultaneously download the same file.  This is a stab at better addressing that problem.  It should rarely occur in practice because most of the data is already on the local file system.

jedwards4b commented:
> Can you elaborate?
I'm not doing it explicitly, but the clients (gftp, ftp, svn, wget) do it internally.

jedwards4b commented:
testing complete - this is ready @billsacks  do you want to review again?
